promql/engine.go

```diff
@@ -852,20 +857,23 @@ func (enh *EvalNodeHelper) dropMetricName(l labels.Labels) labels.Labels {
 	return ret
 }
 
-// signatureFunc is a cached version of signatureFunc.
-func (enh *EvalNodeHelper) signatureFunc(on bool, names ...string) func(labels.Labels) uint64 {
+func yoloString(b []byte) string {
+	return *((*string)(unsafe.Pointer(&b)))
+}
+
+func (enh *EvalNodeHelper) signatureFunc(on bool, names ...string) func(labels.Labels) string {
 	if enh.sigf == nil {
-		enh.sigf = make(map[uint64]uint64, len(enh.out))
+		enh.sigf = make(map[string]string, len(enh.out))
 	}
-	f := signatureFunc(on, names...)
-	return func(l labels.Labels) uint64 {
-		h := l.Hash()
-		ret, ok := enh.sigf[h]
+	f := signatureFunc(on, &enh.lblBuf, names...)
+	return func(l labels.Labels) string {
+		ls := l.YoloString(&enh.lblBuf)
+		ret, ok := enh.sigf[ls]
 		if ok {
 			return ret
 		}
 		ret = f(l)
-		enh.sigf[h] = ret
+		enh.sigf[l.NoRenderString(&enh.lblBuf)] = ret
```

Comment:
You're still calculating this twice, you already have this from `ls` it only needs to be converted to a string.

Reply:
I haven't been able to get reuse of `ls` here to work properly without explicitly copying the byte buffer to a new variable: 
```		
var cp string
copy([]byte(cp), ls)
enh.sigf[cp] = ret
```

Reply:
Hmm, maybe pass around a `[]byte` and only create a Buffer when we need it?